library/std/src/ffi/os_str.rs

```diff
@@ -963,6 +964,82 @@ impl OsStr {
         self.inner.as_encoded_bytes()
     }
 
+    /// Takes a substring based on a range that corresponds to the return value of
+    /// [`OsStr::as_encoded_bytes`].
+    ///
+    /// The range's start and end must lie on valid `OsStr` boundaries.
+    /// A valid `OsStr` boundary is one of:
+    /// - The start of the string
+    /// - The end of the string
+    /// - Immediately before a valid non-empty UTF-8 substring
+    /// - Immediately after a valid non-empty UTF-8 substring
+    ///
+    /// # Panics
+    ///
+    /// Panics if `range` does not lie on valid `OsStr` boundaries or if it
+    /// exceeds the end of the string.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// #![feature(os_str_slice)]
+    ///
+    /// use std::ffi::OsStr;
+    ///
+    /// let os_str = OsStr::new("foo=bar");
+    /// let bytes = os_str.as_encoded_bytes();
+    /// if let Some(index) = bytes.iter().position(|b| *b == b'=') {
+    ///     let key = os_str.slice_encoded_bytes(..index);
+    ///     let value = os_str.slice_encoded_bytes(index + 1..);
+    ///     assert_eq!(key, "foo");
+    ///     assert_eq!(value, "bar");
+    /// }
+    /// ```
+    #[unstable(feature = "os_str_slice", issue = "none")]
+    pub fn slice_encoded_bytes<R: ops::RangeBounds<usize>>(&self, range: R) -> &Self {
+        fn is_valid_boundary(bytes: &[u8], index: usize) -> bool {
+            if index == 0 || index == bytes.len() {
+                return true;
+            }
+
+            // Fast path
+            if bytes[index - 1].is_ascii() || bytes[index].is_ascii() {
+                return true;
+            }
+
+            let (before, after) = bytes.split_at(index);
+
+            // UTF-8 takes at most 4 bytes per codepoint, so we don't
+            // need to check more than that.
+            let after = after.get(..4).unwrap_or(after);
+            match str::from_utf8(after) {
+                Ok(_) => return true,
+                Err(err) if err.valid_up_to() != 0 => return true,
+                Err(_) => (),
+            }
+
+            for len in 2..=4.min(index) {
+                let before = &before[index - len..];
+                if str::from_utf8(before).is_ok() {
+                    return true;
+                }
+            }
+
+            false
```

Comment:
Do we need this?  Shouldn't it be sufficient to check
- First byte is ASCII or UTF-8 start byte
- Byte after last is ASCII or UTF-8 start byte (assuming end-of-buffer is already handled)

Reply:
On Unix anything goes, so we have to prove that there's a full UTF-8 codepoint on either side of both splits. For all we know there's a single incorrect byte in the middle of a codepoint. (Otherwise we could just use reuse the `str` logic.)

On Windows we could get away with doing less work, because it's not just an arbitrary `&[u8]`. Judging by the `wtf8` code it would look something like this for each split:
- Check if the first byte after the split is a start byte/ASCII (if not, fail)
- Check if the first byte is `0xED`/`1110_1101` (if not, succeed)
- Check if the second byte `>= 0xA0`/`1010_0000` (if not, succeed)
- Now we know there's a surrogate after the split, so do something similar to check if there's one before the split as well.